### ISSUE TITLE
Is there any event or callback to know when the DOM has finished updating?

### ISSUE THREAD

**Issue by PierBover** [mikeric/rivets#470]
I'm making a grid of elements and after the DOM has been populated with `rv-each` I need to dynamically set the width and height of the grid elements using jQuery.

How would this be done?

**Comment by Duder-onomy** [comment]
This can be done. But...I hope someone else chimes in with a better solution, cause mine is a little dirty I think.

Check out this fiddle: http://jsfiddle.net/af8qwh8j/

The gist of it is to make a new each binder, I called mine `evented-each-*` and call the regular each binder while wrapping all its methods in your custom events. 

```javascript
rivets.binders['evented-each-*'] = {
    block : true,
    priority : 4000,
    bind : function() {
        $('#fireEventsHere').trigger('bind:begining');
        rivets.binders['each-*'].bind.apply(this, arguments);                       
        $('#fireEventsHere').trigger('bind:complete');
    },
    unbind : function() {
        $('#fireEventsHere').trigger('unbind:begining');
        rivets.binders['each-*'].unbind.apply(this, arguments);
        $('#fireEventsHere').trigger('unbind:complete');
    },
    routine : function() {
        $('#fireEventsHere').trigger('routine:begining');
        rivets.binders['each-*'].routine.apply(this, arguments);
        $('#fireEventsHere').trigger('routine:complete');
    },
    update : function() {
        $('#fireEventsHere').trigger('update:begining');
        rivets.binders['each-*'].update.apply(this, arguments);
        $('#fireEventsHere').trigger('update:complete');
    }
};
```

As you remove names from the list, you will see new events fire. Pretty cool stuff.

Hope this helps.

**Comment by PierBover** [comment]
Thanks again for your help!

When does the update callback get triggered?

**Comment by Duder-onomy** [comment]
It is an internal method inside the [`each-*` binder](https://github.com/mikeric/rivets/blob/master/src/binders.coffee#L218). 

It looks like it gets called when, after initially looping though the collection and building the initial DOM, one of the items in the collections model changes.

**Comment by PierBover** [comment]
It would be great if those callbacks or events were included by default on Rivets.

Can you mark it as a feature request?

**Comment by Duder-onomy** [comment]
@PierBover I am not a collab on this repo, I cannot mark stuff as bug etc.  
Though, from lurking on this Repo pretty hard for the last year or so, my feeling is that the community would rather you just write your own binders.  If you really feel strongly about it you should ping `mikeric`.

I am currently working on a `each-*` binder that will have full event support bove and will also keep the DOM in order. It will fire `beforeEnter` `afterEnter` and `beforeExit` events on the iterated models so you can easily animate the iterated items in and out of the DOM like, 

```html
<ul>
    <li rv-each-item='model.collection' rv-on-beforeEnter='set height to zero' rv-on-afterEnter='animate into DOM ' rv-on-beforeExit='animate out of DOM' ></li>
</ul>
```

The only problem with this right now is that Rivets does not keep the actual DOM elements in order. 

For example, if you splice the center item out of an array of 3 items that is bound via `each-*`. You would expect it to just remove the middle element from the DOM. But, it actually removes the last item from the DOM and then re-binds the second item in the DOM with the model of the previous third item in the collection.  Anyhow, I will keep a note to Ping you when I finish my `each-*` binder. It sounds like it could be useful to you.

**Comment by PierBover** [comment]
This is interesting. Yes please, ping me when it's finished.

**Comment by NV0** [comment]
Agree, I think there should be some way to get notified when a model get bound and rendered, so further DOM manipulations can be made without hacks (setInterval or something).

**Comment by PierBover** [comment]
Can someone add the "feature request" label, please?

**Comment by Duder-onomy** [comment]
@NVO and @PierBover  Here is a [previous issue thread](https://github.com/mikeric/rivets/issues/337) talking about something similar. 
Ammo for your arsenal?

I personally feel that (since rivets does not currently send any special events) that maybe you should make a rivets plugin? `evented-rivets` or something like that. Then you can monkey patch events into all the binders you want?  In the apps I work on, I am usually only concerned about when the items are entering and exiting the DOM. `rv-if` `rv-show` etc. So that I might animate them without attaching animation logic to the model informing the DOM.

I believe `rivets.bind` is sync. So you know rivets is done binding then that method is finished.

What exactly are you looking for rivets to support? just `on-bind` or other events as well?

**Comment by PierBover** [comment]
> What exactly are you looking for rivets to support? just on-bind or other events as well?

I'm new to Rivets so I can't really say a general case, but my particular problem was that I needed a callback whenever some piece of the DOM was updated by Rivets.

What I ended up doing any time I changed the model was to `unbind()` Rivets, update the objects Rivets is reading, `bind()`, and then do my DOM stuff.

**Comment by davekiss** [comment]
:+1:

**Comment by ksikka** [comment]
+1 for a callback that gets called after some piece of the DOM is updated by rivets. It would allow our parent views to listen for subview changes and potentially run DOM post-processing across the entire view.

Example use case: say you want to make all inputs and buttons have disabled = true if some top level view attribute called "readOnly" is true, but these inputs and buttons are getting rendered by subviews. To solve this, you could observe rivets to know when a subview changed, and then apply disable=true to all DOM elements in the DOM subtree of the parent view.

**Comment by kojilab** [comment]
Looking for the same thing but on top of a component.

**Comment by kojilab** [comment]
Thanks to Duder-onomy for the example. Helped me build a custom rv-each with a ready callback
```

rivets.binders['each-*-onready-*'] = {
	block : true,
    priority : 4000,
    bind : function() {        
        rivets.binders['each-*'].bind.apply(this, arguments);                               
    },
    unbind : function() {        
        rivets.binders['each-*'].unbind.apply(this, arguments);        
    },
    routine : function(el, collection) {        
        rivets.binders['each-*'].routine.apply(this, arguments);
		var callbackName = this.args[1].replace(/-([a-z])/g, function(grouped) {
        	return grouped[1].toUpperCase();
      	});		
		if (_.isArray(this.iterated) && this.iterated.length === collection.length) {						
			this.view.models[callbackName]();
		}     
    },
    update : function() {        
        rivets.binders['each-*'].update.apply(this, arguments);        
    }
};
```

**Comment by JumpLink** [comment]
You could try this:

```js
rivets.components['your-component'] = {
  template: function() {
    [...]
  },
  initialize: function(el, data) {
    var controller = this;  
    var $el = $(el);

    var ready = function() {
      // do your stuff after view is rendered here
    }

    $el.bind('DOMSubtreeModified', ready);
    return controller;
  }
};
```
